cozy/codegen/cxx.py

```diff
@@ -274,6 +274,14 @@ def construct_concrete(self, t : Type, e : Exp, out : Exp):
             return SEscape("{indent}{lhs} = {rhs};\n", ["lhs", "rhs"], [out, e])
         elif is_scalar(t):
             return SEscape("{indent}{lhs} = {rhs};\n", ["lhs", "rhs"], [out, e])
+        elif isinstance(t, TTuple):
+            x = self.fv(t, "x")
+            indices = range(len(t.ts))
+            return SSeq(
+                SEscape("{indent}{rv} = {rhs};\n", ["rv", "rhs"], [x, e]),
+                SEscape("{indent}{lhs} = {rv};\n", ["lhs", "rv"],
+                        [out, ETuple(tuple(map(lambda i : ETupleGet(x, i), indices))).with_type(t)]),
```

Comment:
I suspect this code will crash unless the `ETuple` and `ETupleGet` expressions are annotated with correct types using `.with_type(...)`.

Also, a minor note: Python programmers generally prefer list comprehensions to functional maps and folds. Your use of `map` could be written `[ETupleGet(x, i) for i in range(len(t.ts))]` instead. In Cozy this is an especially good idea since there are many variables named `map` floating around that can shadow the builtin function.

Reply:
Thanks! updated